enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,343 @@
+replaces: []
+superseded-by: []
+---
+
+# Pin and pre-load images
+
+## Summary
+
+Provide an mechanism to pin and pre-load container images.
+
+## Motivation
+
+Slow and/or unreliable connections to the image registry servers interfere with
+operations that require pulling images. For example, an upgrade may require
+pulling more than one hundred images. Failures to pull those images cause
+retries that interfere with the upgrade process and may eventually make it
+fail. One way to improve that is to pull the images in advance, before they are
+actually needed, and ensure that they aren't removed. Doing that provides a
+more consistent upgrade time in those environments. That is important when
+scheduling upgrades into maintenance windows, even if the upgrade might not
+otherwise fail.
+
+### User Stories
+
+#### Pre-load and pin upgrade images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load all the
+images required for the upgrade in advance, so that when I decide to actually
+perform the upgrade there will be no need to contact that slow and/or
+unreliable registry server and the upgrade will successfully complete in a
+predictable time.
+
+#### Pre-load and pin application images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load the images
+required by my application in advance, so that when I decide to actually deploy
+it there will be no need to contact that slow and/or unreliable registry server
+and my application will successfully deploy in a predictable time.
+
+### Goals
+
+Provide a mechanism that cluster administrators can use to pin and pre-load
+container images.
+
+### Non-Goals
+
+We wish to use the mechanism described in this enhancement to orchestrate
+upgrades without a registry server. But that orchestration is not a goal
+of this enhancement; it will be part of a separate enhancement based on
+parts of https://github.com/openshift/enhancements/pull/1432.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster uses the new `PinnedImageSet` object to
+request that a set of container images are pinned and pre-loaded:
+
+    ```yaml
+    apiVersion: machineconfiguration.openshift.io/v1alpha1
+    kind: PinnedImageSet
+    metadata:
+      name: my-pinned-images
+    spec:
+      nodeSelector:
+        matchLabels:
+          node-role.kubernetes.io/control-plane: ""
+      pinnedImages:
+      - quay.io/openshift-release-dev/ocp-release@sha256:...
+      - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+      - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+      ...
+    ```
+
+1. The machine config operators ensures that all the images are pinned and
+pulled in all the nodes that match the node selector.
+
+### API Extensions
```

Comment:
As per my understanding, we will also be introduce a pinnedImageSet CRD in MCO?

Reply:
Yes, that is what I mention below:

> A new `PinnedImageSet` object will be added to the
> `machineconfiguration.openshift.io` API group.

Am I missing something?

Reply:
Yes, that is what I mean when I say "object". I replaced "object" with "custom resource" to try to make it clear. So, we will have a new `PinnedImageSet` custom resource. It is defined in detail in https://github.com/openshift/api/pull/1609.

Reply:
For clarity, a new type in the k8s API is a CRD (Custom Resource Definition) and an instance of that type is a CR (or object)</pedantry>